sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -157,7 +157,7 @@ abstract class Optimizer(catalogManager: CatalogManager)
     //   since the other rules might make two separate Unions operators adjacent.
     Batch("Inline CTE", Once,
       InlineCTE()) ::
-    Batch("Union", Once,
+    Batch("Union", fixedPoint,
```

Comment:
Why change to fixed point?

Reply:
Without the pre-optimization in the dataframe side, some pyspark tests failed complaining that this batch is not idempotent. In fact, this optimization batch does not need to be idempotent, we use `Once` before because we thought it was sufficient, but it's actually not sufficient.